pkg/ovs/openflow/ofctrl_action.go

```diff
@@ -83,59 +83,80 @@ func (a *ofFlowAction) CT(commit bool, tableID uint8, zone int, zoneSrcField *Re
 // ofCTAction is a struct to implement CTAction.
 type ofCTAction struct {
 	ctBase
-	actions []openflow13.Action
+	actions []openflow15.Action
 	builder *ofFlowBuilder
 }
 
 // LoadToMark is an action to load data into ct_mark.
 func (a *ofCTAction) LoadToMark(value uint32) CTAction {
-	field, rng, _ := getFieldRange(NxmFieldCtMark)
-	a.load(field, uint64(value), &rng)
+	ctMarkField := openflow15.NewCTMarkMatchField(value, nil)
+	action := openflow15.NewActionSetField(*ctMarkField)
+	a.actions = append(a.actions, action)
 	return a
 }
 
 func (a *ofCTAction) LoadToCtMark(marks ...*CtMark) CTAction {
-	field, _, _ := getFieldRange(NxmFieldCtMark)
 	for _, mark := range marks {
-		a.load(field, uint64(mark.value), mark.field.rng)
+		var mask *uint32
+		maskData := uint32(0)
+		valueData := mark.value
+		if mark.field.rng != nil {
+			maskData = ^maskData >> (32 - mark.field.rng.Length()) << mark.field.rng.Offset()
+			mask = &maskData
+			valueData = valueData << mark.field.rng.Offset()
+		}
+		ctMarkField := openflow15.NewCTMarkMatchField(valueData, mask)
+		action := openflow15.NewActionSetField(*ctMarkField)
+		a.actions = append(a.actions, action)
 	}
 	return a
 }
 
 func (a *ofCTAction) LoadToLabelField(value uint64, labelField *CtLabel) CTAction {
-	field, _, _ := getFieldRange(NxmFieldCtLabel)
-	a.load(field, value, labelField.rng)
-	return a
-}
-
-func (a *ofCTAction) load(field *openflow13.MatchField, value uint64, rng *Range) {
-	action := openflow13.NewNXActionRegLoad(rng.ToNXRange().ToOfsBits(), field, value)
+	var labelBytes, maskBytes [16]byte
+	valueData := value
+	if labelField.rng != nil {
+		mask := ^uint64(0) >> (64 - labelField.rng.Length()) << labelField.rng.Offset()
+		valueData = valueData << (labelField.rng.Offset() % 64)
+		if labelField.rng.Offset() > 64 {
+			binary.BigEndian.PutUint64(maskBytes[0:8], mask)
+			binary.BigEndian.PutUint64(labelBytes[0:8], valueData)
+		} else {
+			binary.BigEndian.PutUint64(maskBytes[8:], mask)
+			binary.BigEndian.PutUint64(labelBytes[8:], valueData)
+		}
+	} else {
+		binary.BigEndian.PutUint64(labelBytes[0:8], valueData)
+	}
+	ctLabelField := openflow15.NewCTLabelMatchField(labelBytes, &maskBytes)
+	action := openflow15.NewActionSetField(*ctLabelField)
 	a.actions = append(a.actions, action)
+	return a
 }
 
 // MoveToLabel is an action to move data into ct_label.
 func (a *ofCTAction) MoveToLabel(fromName string, fromRng, labelRng *Range) CTAction {
-	fromField, _ := openflow13.FindFieldHeaderByName(fromName, false)
-	toField, _ := openflow13.FindFieldHeaderByName(NxmFieldCtLabel, false)
+	fromField, _ := openflow15.FindOxmIdByName(fromName, false)
```

Comment:
Why `FindOxmIdByName`? I saw in some places `FindFieldHeaderByName` is also used.

Reply:
The return types of `FindOxmIdByName` and `FindFieldHeaderByName` are different. `FindFieldHeaderByName` returns "MatchField", and `FindOxmIdByName` returns "OxmId". "OxmId" is newly introduced in libOpenflow for OpenFlow1.5, and it is used in new openflow action `CopyFieldAction` in OF1.5